fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/RecordMetaDataBuilder.java

```diff
@@ -484,111 +576,116 @@ private static void validateRecords(@Nonnull Descriptors.FileDescriptor fileDesc
                                             " should not be repeated");
             }
-            Descriptors.Descriptor descriptor = unionField.getMessageType();
+            Descriptors.Descriptor descriptor = adjustedUnionField.getMessageType();
             if (!unionFields.containsKey(descriptor)) {
-                RecordMetaDataOptionsProto.RecordTypeOptions recordTypeOptions = descriptor.getOptions()
-                        .getExtension(RecordMetaDataOptionsProto.record);
-                if (recordTypeOptions != null &&
-                        recordTypeOptions.getUsage() != RecordMetaDataOptionsProto.RecordTypeOptions.Usage.RECORD) {
-                    throw new MetaDataException("Union field " + unionField.getName() +
-                                                " has type " + descriptor.getName() +
-                                                " which is not a record");
-                }
-
-                if (descriptor.getFile() != recordsDescriptor) {
-                    // An imported record type.
-                    RecordTypeBuilder recordType = new RecordTypeBuilder(descriptor);
-                    if (recordTypes.putIfAbsent(recordType.getName(), recordType) != null) {
-                        throw new MetaDataException("There is already a record type named" + recordType.getName());
-                    }
-                    if (processExtensionOptions) {
-                        protoFieldOptions(recordType, descriptor);
-                    }
-                }
-
-                unionFields.put(descriptor, unionField);
+                processRecordType(adjustedUnionField, processExtensionOptions);
+                unionFields.put(descriptor, adjustedUnionField);
             } else {
                 // The preferred field is the last one, except if there is one whose name matches.
-                unionFields.compute(descriptor, (d, f) -> f != null && f.getName().equals("_" + d.getName()) ? f : unionField);
+                unionFields.compute(descriptor, (d, f) -> f != null && f.getName().equals("_" + d.getName()) ? f : adjustedUnionField);
             }
         }
     }
 
-    private void protoFieldOptions(RecordTypeBuilder recordType, Descriptors.Descriptor descriptor) {
+    private void processRecordType(@Nonnull Descriptors.FieldDescriptor unionField, boolean processExtensionOptions) {
+        Descriptors.Descriptor descriptor = unionField.getMessageType();
+        RecordMetaDataOptionsProto.RecordTypeOptions recordTypeOptions = descriptor.getOptions()
+                .getExtension(RecordMetaDataOptionsProto.record);
+        if (recordTypeOptions != null &&
+            recordTypeOptions.getUsage() != RecordMetaDataOptionsProto.RecordTypeOptions.Usage.RECORD) {
+            throw new MetaDataException("Union field " + unionField.getName() +
+                                        " has type " + descriptor.getName() +
+                                        " which is not a record");
```

Comment:
right. Also to clarify: the validation done in validateUnion is the other way around. It verifies that all RECORD message types defined in this proto are also present in the union. As you also mentioned, here we are checking if a union member is a RECORD message type, which effectively double checks the message types that are present in the file in addition to checking the message types that are imported. I think it may be cleaner and clearer to move this to validateUnion too. I'll add a comment and put it there. Thanks.

Reply:
Doesn't it validate both ways in `validateUnion`? It validates that all `RECORD`s are in the union (after the `switch` statement currently) and that all `NESTED`s and `UNION`s are not in the union in their respective `switch` cases, if I'm reading it correctly.

I'm not sure `validateUnion` is the right place to put this (or, like, I'm not sure how I see how it's *better* there than here) if the idea is that it should be validating imported records as well.

Reply:
The first part is right. My point was: validateUnion only validates the message types defined in the original records descriptor. It does not check if ALL of the union fields (i.e., the imported message types) are RECORD types. This check is meant to do that. Moving it validateUnion makes sense, IMO, because it puts all major union-related validations together.

Reply:
Oh, I see; I think I misunderstood your comment.

I could see putting this check here or in `validateUnion` going either way in terms of cleanliness. Your call, I suppose.